// frontend/src/lib/theme.ts
export const theme = {
    colors: {
        primary: {
            DEFAULT: '#4f46e5', // indigo-600
            light: '#6366f1',    // indigo-500
            dark: '#4338ca',     // indigo-700
            foreground: '#ffffff'
        },
        secondary: {
            DEFAULT: '#10b981', // emerald-500
            light: '#34d399',   // emerald-400
            dark: '#059669',    // emerald-600
            foreground: '#ffffff'
        },
        accent: {
            DEFAULT: '#f59e0b', // amber-500
            light: '#fbbf24',   // amber-400
            foreground: '#ffffff'
        },
        danger: {
            DEFAULT: '#ef4444', // red-500
            light: '#f87171',   // red-400
            foreground: '#ffffff'
        },
        success: {
            DEFAULT: '#10b981', // emerald-500
            light: '#34d399',   // emerald-400
            foreground: '#ffffff'
        },
        warning: {
            DEFAULT: '#f59e0b', // amber-500
            light: '#fbbf24',   // amber-400
            foreground: '#ffffff'
        },
        background: '#f9fafb', // gray-50
        card: '#ffffff',
        text: {
            primary: '#111827', // gray-900
            secondary: '#6b7280', // gray-500
            muted: '#9ca3af'      // gray-400
        }
    },
    borderRadius: {
        sm: '0.375rem',
        md: '0.5rem',
        lg: '0.75rem',
        xl: '1rem',    // 16px
        '2xl': '1.5rem'
    },
    shadows: {
        sm: '0 1px 2px 0 rgb(0 0 0 / 0.05)',
        md: '0 4px 6px -1px rgb(0 0 0 / 0.1), 0 2px 4px -2px rgb(0 0 0 / 0.1)',
        lg: '0 10px 15px -3px rgb(0 0 0 / 0.1), 0 4px 6px -4px rgb(0 0 0 / 0.1)',
    }
};
