// frontend/src/components/dashboard/GlobalStudentModal.tsx
import React, { useState, useEffect } from 'react';
import {
    Dialog,
    DialogContent,
    DialogHeader,
    DialogTitle,
    DialogDescription,
} from '@/components/ui/dialog';
import { useQuery, useQueryClient } from '@tanstack/react-query';
import api from '@/services/api';
import { DataTable } from './DataTable';
import { studentColumns } from './StudentListColumns';
import { AnalyticsFilters } from '@/hooks/useAnalytics';
import { Search, Users } from 'lucide-react';
import { toast } from 'sonner';
import { Input } from '@/components/ui/input';
import { useAuth } from '@/context/AuthContext';
import { ReasonModal } from './ReasonModal';

interface GlobalStudentModalProps {
    isOpen: boolean;
    onClose: () => void;
    filters: AnalyticsFilters;
}

const GlobalStudentModal: React.FC<GlobalStudentModalProps> = ({ isOpen, onClose, filters }) => {
    const { user } = useAuth();
    const queryClient = useQueryClient();
    const [search, setSearch] = useState('');
    const [debouncedSearch, setDebouncedSearch] = useState('');

    // Modal State
    const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false);
    const [isRejectModalOpen, setIsRejectModalOpen] = useState(false);
    const [activityToProcess, setActivityToProcess] = useState<any>(null);

    // Debounce Search
    useEffect(() => {
        const timer = setTimeout(() => {
            setDebouncedSearch(search);
        }, 500);
        return () => clearTimeout(timer);
    }, [search]);

    // Fetch Global Student List
    const { data: students, isLoading } = useQuery({
        queryKey: ['global-students', filters, debouncedSearch],
        queryFn: async () => {
            const { data } = await api.get('/analytics/student-list', {
                params: {
                    ...filters,
                    search: debouncedSearch,
                    per_page: 100
                }
            });
            // API returns { students: [], total_pages: ... }
            return data.data.students || [];
        },
        enabled: isOpen,
    });

    const handleDeleteActivity = (activity: any) => {
        setActivityToProcess(activity);
        setIsDeleteModalOpen(true);
    };

    const handleRejectActivity = (activity: any) => {
        setActivityToProcess(activity);
        setIsRejectModalOpen(true);
    };

    const confirmDeleteActivity = async (reason: string) => {
        if (!activityToProcess) return;
        try {
            await api.post(`/admin/student-activities/delete/${activityToProcess.id}`, { reason });
            queryClient.invalidateQueries({ queryKey: ['global-students'] });
            queryClient.invalidateQueries({ queryKey: ['students'] });
            queryClient.invalidateQueries({ queryKey: ['kpi-summary'] });
            toast.success("Activity deleted.");
        } catch (error: any) {
            toast.error(error.error || "Deletion failed");
            throw error;
        }
    };

    const confirmRejectActivity = async (reason: string) => {
        if (!activityToProcess) return;
        try {
            // Use the faculty reject endpoint which supports reason
            await api.post(`/faculty/reject/${activityToProcess.id}`, { faculty_comment: reason });
            queryClient.invalidateQueries({ queryKey: ['global-students'] });
            queryClient.invalidateQueries({ queryKey: ['students'] });
            queryClient.invalidateQueries({ queryKey: ['kpi-summary'] });
            toast.success("Approval undone. Activity rejected.");
        } catch (error: any) {
            toast.error(error.error || "Rejection failed");
            throw error;
        }
    };

    return (
        <Dialog open={isOpen} onOpenChange={(open) => !open && onClose()}>
            <DialogContent className="max-w-5xl max-h-[90vh] overflow-hidden flex flex-col p-0 rounded-2xl border-none shadow-2xl gap-0">
                <DialogHeader className="p-8 pb-6 bg-gradient-to-r from-slate-50 to-white dark:from-slate-900 dark:to-slate-950 space-y-6">
                    <div className="flex items-center justify-between">
                        <div className="space-y-1">
                            <div className="inline-flex items-center gap-2 px-2.5 py-0.5 rounded-full bg-indigo-50 dark:bg-indigo-900/20 text-indigo-600 dark:text-indigo-400 text-[10px] font-bold uppercase tracking-wider border border-indigo-100 dark:border-indigo-900/30">
                                <Users className="h-3 w-3" />
                                <span>Master Database</span>
                            </div>
                            <DialogTitle className="text-3xl font-extrabold tracking-tight text-slate-900 dark:text-white">All Records</DialogTitle>
                            <DialogDescription className="text-slate-500 font-medium">
                                Direct administrative access to the global student activity catalog.
                            </DialogDescription>
                        </div>
                    </div>

                    <div className="relative group">
                        <Search className="absolute left-4 top-1/2 -translate-y-1/2 h-4 w-4 text-slate-400 group-focus-within:text-indigo-500 transition-colors" />
                        <Input
                            placeholder="Universal search by Name, ID, or Event..."
                            className="pl-11 h-12 bg-white dark:bg-slate-900 border-slate-200 dark:border-slate-800 rounded-xl shadow-sm focus:ring-indigo-500/10 placeholder:text-slate-400 transition-all font-medium"
                            value={search}
                            onChange={(e) => setSearch(e.target.value)}
                        />
                    </div>
                </DialogHeader>

                <div className="flex-1 overflow-y-auto p-8 pt-2 bg-white dark:bg-slate-950">
                    <div className="rounded-2xl border border-slate-100 dark:border-slate-800 shadow-sm overflow-hidden bg-slate-50/30 dark:bg-slate-900/10">
                        <DataTable
                            columns={studentColumns}
                            data={students || []}
                            loading={isLoading}
                            pageCount={1}
                            pagination={{ pageIndex: 0, pageSize: 20 }}
                            onPaginationChange={() => { }}
                            sorting={[]}
                            onSortingChange={() => { }}
                            options={{
                                meta: {
                                    onDelete: (user?.role?.toLowerCase() === 'admin' || user?.role?.toLowerCase() === 'faculty') ? handleDeleteActivity : undefined,
                                    onReject: (user?.role?.toLowerCase() === 'admin' || user?.role?.toLowerCase() === 'faculty') ? handleRejectActivity : undefined,
                                    user
                                }
                            }}
                        />
                    </div>
                </div>

                <ReasonModal
                    isOpen={isDeleteModalOpen}
                    onClose={() => {
                        setIsDeleteModalOpen(false);
                        setActivityToProcess(null);
                    }}
                    onConfirm={confirmDeleteActivity}
                    title="Delete Activity Record"
                    description={`Are you sure you want to delete the activity "${activityToProcess?.title}" for student "${activityToProcess?.student_name}"? This action cannot be undone.`}
                    variant="destructive"
                    icon="delete"
                    confirmLabel="Delete Permanently"
                    placeholder="Provide a reason for this deletion (e.g. Fraudulent submission, duplicate entry)..."
                />

                <ReasonModal
                    isOpen={isRejectModalOpen}
                    onClose={() => {
                        setIsRejectModalOpen(false);
                        setActivityToProcess(null);
                    }}
                    onConfirm={confirmRejectActivity}
                    title="Undo Approval / Reject"
                    description={`This will reject the certificate for "${activityToProcess?.student_name}". The student will be notified to re-upload.`}
                    variant="warning"
                    icon="reject"
                    confirmLabel="Reject & Notify"
                    placeholder="Provide a clear reason (e.g. Blurry photo, wrong event mentioned)..."
                />
            </DialogContent>
        </Dialog>
    );
};

export default GlobalStudentModal;
