// frontend/src/components/dashboard/ReasonModal.tsx
import React, { useState } from 'react';
import { cn } from '@/lib/utils';
import {
    Dialog,
    DialogContent,
    DialogHeader,
    DialogTitle,
    DialogDescription,
    DialogFooter,
} from '@/components/ui/dialog';
import { Button } from '@/components/ui/button';
import { Label } from '@/components/ui/label';
import { Textarea } from '@/components/ui/textarea';
import { AlertTriangle, XCircle, Trash2, Loader2 } from 'lucide-react';

interface ReasonModalProps {
    isOpen: boolean;
    onClose: () => void;
    onConfirm: (reason: string) => Promise<void>;
    title: string;
    description: string;
    placeholder?: string;
    confirmLabel?: string;
    variant?: 'destructive' | 'warning' | 'default';
    icon?: 'delete' | 'reject' | 'info';
}

type ButtonVariant = "default" | "destructive" | "outline" | "secondary" | "ghost" | "link";

export const ReasonModal: React.FC<ReasonModalProps> = ({
    isOpen,
    onClose,
    onConfirm,
    title,
    description,
    placeholder = "Type your reason here...",
    confirmLabel = "Confirm Action",
    variant = 'default',
    icon = 'info'
}) => {
    const [reason, setReason] = useState('');
    const [loading, setLoading] = useState(false);

    const handleConfirm = async () => {
        if (!reason.trim()) return;
        setLoading(true);
        try {
            await onConfirm(reason);
            setReason('');
            onClose();
        } catch (error) {
            console.error("Action failed", error);
        } finally {
            setLoading(false);
        }
    };

    const getIcon = () => {
        switch (icon) {
            case 'delete': return <Trash2 className="h-5 w-5 text-rose-500" />;
            case 'reject': return <XCircle className="h-5 w-5 text-amber-500" />;
            default: return <AlertTriangle className="h-5 w-5 text-indigo-500" />;
        }
    };

    // Map custom variants to valid Button variants
    const getButtonVariant = (): ButtonVariant => {
        if (variant === 'destructive') return 'destructive';
        if (variant === 'warning') return 'secondary';
        return 'default';
    };

    return (
        <Dialog open={isOpen} onOpenChange={(open) => !open && !loading && onClose()}>
            <DialogContent className="sm:max-w-[425px] overflow-hidden border-none shadow-2xl p-0">
                <div className="bg-white dark:bg-slate-950 p-6">
                    <DialogHeader className="mb-6">
                        <div className="flex items-center gap-4">
                            <div className={`p-3 rounded-2xl shadow-sm ${
                                variant === 'destructive' ? 'bg-rose-50 text-rose-600 dark:bg-rose-900/20' : 
                                variant === 'warning' ? 'bg-amber-50 text-amber-600 dark:bg-amber-900/20' : 
                                'bg-indigo-50 text-indigo-600 dark:bg-indigo-900/20'
                            }`}>
                                {getIcon()}
                            </div>
                            <div className="space-y-1">
                                <DialogTitle className="text-xl font-bold tracking-tight">{title}</DialogTitle>
                                <DialogDescription className="text-slate-500 text-sm leading-relaxed">
                                    {description}
                                </DialogDescription>
                            </div>
                        </div>
                    </DialogHeader>

                    <div className="space-y-4 pt-2">
                        <div className="space-y-2">
                            <Label htmlFor="reason" className="text-[10px] uppercase tracking-[0.1em] font-bold text-slate-400">
                                Required Feedback
                            </Label>
                            <Textarea
                                id="reason"
                                placeholder={placeholder}
                                value={reason}
                                onChange={(e) => setReason(e.target.value)}
                                className="min-h-[120px] bg-slate-50/50 dark:bg-slate-900/50 border-slate-200 dark:border-slate-800 rounded-xl focus:ring-indigo-500/20 resize-none text-sm p-4 placeholder:text-slate-400"
                                autoFocus
                            />
                        </div>
                    </div>
                </div>

                <DialogFooter className="bg-slate-50/50 dark:bg-slate-900/50 px-6 py-4 border-t border-slate-100 dark:border-slate-800 flex items-center gap-3">
                    <Button 
                        variant="ghost" 
                        onClick={onClose} 
                        disabled={loading}
                        className="rounded-lg font-medium text-slate-500 hover:text-slate-900"
                    >
                        Cancel
                    </Button>
                    <Button 
                        variant={getButtonVariant()} 
                        onClick={handleConfirm} 
                        disabled={loading || !reason.trim()}
                        className={cn(
                            "rounded-lg font-bold min-w-[140px] shadow-sm transition-all",
                            variant === 'warning' && "bg-amber-500 hover:bg-amber-600 text-white border-none shadow-amber-200 dark:shadow-none",
                            variant === 'destructive' && "bg-rose-600 hover:bg-rose-700 shadow-rose-200 dark:shadow-none",
                            variant === 'default' && "bg-indigo-600 hover:bg-indigo-700 shadow-indigo-200 dark:shadow-none"
                        )}
                    >
                        {loading ? <Loader2 className="h-4 w-4 animate-spin mr-2" /> : null}
                        {confirmLabel}
                    </Button>
                </DialogFooter>
            </DialogContent>
        </Dialog>
    );
};
